File: second.swift
import Foundation

func findFinalDestination(_ routes: [(String, String)]) -> String {
    if routes.isEmpty {
        return "обратитесь к специалисту"
    }
    
    var routeMap = [String: String]()
    var isDestination = [String: Bool]()
    
    for route in routes {
        let from = route.0
        let to = route.1
        
        if from == to {
            return "обратитесь к специалисту"
        }
        
        routeMap[from] = to
        isDestination[to] = true
    }
    
    var startPoint: String?
    for route in routes {
        if isDestination[route.0] == nil {
            startPoint = route.0
            break
        }
    }
    
    guard let start = startPoint else {
        return "обратитесь к специалисту"
    }
    
    var current = start
    var longestPathEnd = current
    
    while let next = routeMap[current] {
        current = next
        longestPathEnd = current
    }
    
    return longestPathEnd
}

func main() {
    var routes = [(String, String)]()
    
    print("Введите количество пар маршрутов: ", terminator: "")
    guard let n = Int(readLine() ?? "") else {
        print("Некорректный ввод")
        return
    }
    
    for i in 0..<n {
        print("Введите начальный пункт \(i + 1): ", terminator: "")
        let from = readLine() ?? ""
        print("Введите конечный пункт \(i + 1): ", terminator: "")
        let to = readLine() ?? ""
        routes.append((from, to))
    }
    
    let result = findFinalDestination(routes)
    print("Итоговый конечный пункт самой длинной цепочки: \(result)")
}

// Вызов основной функции
main()

File: second.rs
use std::collections::HashMap;
use std::io;

fn find_final_destination(routes: Vec<(String, String)>) -> String {
    if routes.is_empty() {
        return "обратитесь к специалисту".to_string();
    }

    let mut route_map = HashMap::new();
    let mut is_destination = HashMap::new();

    for (from, to) in &routes {
        if from == to {
            return "обратитесь к специалисту".to_string();
        }

        route_map.insert(from.clone(), to.clone());
        is_destination.insert(to.clone(), true);
    }

    let start_point = routes.iter()
        .find(|(from, _)| !is_destination.contains_key(from))
        .map(|(from, _)| from.clone());

    let start_point = match start_point {
        Some(point) => point,
        None => return "обратитесь к специалисту".to_string(),
    };

    let mut current = start_point;
    let mut longest_path_end = current.clone();

    while let Some(next) = route_map.get(&current) {
        current = next.clone();
        longest_path_end = current.clone();
    }

    longest_path_end
}

fn main() {
    let mut routes = Vec::new();
    let mut n = String::new();

    println!("Введите количество пар маршрутов: ");
    io::stdin().read_line(&mut n).expect("Ошибка чтения");
    let n: usize = n.trim().parse().expect("Введите число");

    for i in 0..n {
        let mut from = String::new();
        let mut to = String::new();

        println!("Введите начальный пункт {}: ", i + 1);
        io::stdin().read_line(&mut from).expect("Ошибка чтения");
        let from = from.trim().to_string();

        println!("Введите конечный пункт {}: ", i + 1);
        io::stdin().read_line(&mut to).expect("Ошибка чтения");
        let to = to.trim().to_string();

        routes.push((from, to));
    }

    let result = find_final_destination(routes);
    println!("Итоговый конечный пункт самой длинной цепочки: {}", result);
}